quickstarts/WFLY-15746_WildFly_Application_Development_Guide.adoc

```diff
@@ -0,0 +1,83 @@
+= WildFly Application Development Guide
+:author:            Eduardo Martins
+:email:             emartins@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This document specifies the development of the Development Guide feature enhancement.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-15746[WFLY-15746]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1845[EAP7-1845]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:
+
+=== Affected Projects or Components
+
+WildFly (WFLY Documentation and Quickstarts Components)
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* The guide should have 2 major sections, one for first time/beginners app developers, and one for advanced app developers.
+
+* The major section for first time app developers should include the following new content:
+** How to develop a first basic application (helloworld quickstart) from scratch using Maven, a complete user story walkthrough, which should include:
+*** How to create the application Maven Project, and add Red Hat Maven repositories and BOMs dependencies
```

Comment:
Are you planning to leverage https://issues.redhat.com/browse/WFLY-17651 for these instructions?

With https://issues.redhat.com/browse/WFLY-18428, we want to have a user being able to get up and running with WildFly in a few minutes. This is achievable by using a Maven archetypes that will do all this plumbing for them.

Reply:
No, the idea was to start with the standard Maven archetypes, and document the changes needed. The main reason for this is that WildFly archetypes are not supported, but I also thought of aligning with Quarkus, and reuse their enhancements on app development, both the online tool and the cli app project generation by maven plugin.

Reply:
> WildFly archetypes are not supported

We have been providing them and maintaining them. I don't see why we could not use them to bootstrap the necessary Maven project.
I can do a `mvn archetype:generate` and starts writing *my* code in a matter of minutes. Why would I start from scratch and spend 1/2 to 1 hour just to have the pieces in place in the `pom.xm`, `src/main` & `src/test`?

Reply:
> We have been providing them and maintaining them. I don't see why we could not use them to bootstrap the necessary Maven project. 

Because that means we need to make archetypes a product and support it for 10 years, please remember this is the upstream for a new EAP 8 requirement.

> I can do a mvn archetype:generate and starts writing my code in a matter of minutes. Why would I start from scratch and spend 1/2 to 1 hour just to have the pieces in place in the pom.xm, src/main & src/test?

Nobody wants to start from scratch, for now use the standard Maven archetypes, next use whatever is best in the long term, if possible aligned with Quarkus (which AFAIK opted for a Maven Plugin instead of archetypes).

Reply:
Please note that this doesn't mean we can not add WildFly Archetypes in the upstream guide too, we just need to not make the guide work around it since it's not supported. For instance we can document the supported way and then add a section introducing WFLY Archetypes, and how it automates what we first did manually (add plugin and dep management, import boms, etc.)

Reply:
Archetypes existed for ages, and it is not and never was a popular Maven feature among app developers, mostly cause content is not flexible enough and gets deprecated fast, are you sure it's the best direction for us to promote long term?

Anyway let's say it takes 30 min for the user to manually add everything to the Maven Project... isn't that only for the 1st time for the user to learn about our plugins, BOMs, etc? And won't that learning be needed? And would actually those few minutes really mean something on the time it takes to develop a new app? I can understand the need for an as fast as possible "server start", for an as fast as possible "get started", now an as fast as possible "app development"? I can't really see what's wrong with now explaining how to do it manually, starting with the maven archetypes + (upstream only) extend that with the "unstable/unsupported" features like Archetypes.

Reply:
@ehsavoie @jmesnil This sounds like something that can be handled via doc modularization. I see the value in a content section that spells things out, and also in one that tells you how to use things WildFly provides to do that for you. The guide incorporates both with a bit of preface to tell you about and link to the two options.

The endpoint for both paths would be a consistent maven project.

The "how to use things WildFly provides" cannot be constrained by EAP. WildFly will always be ahead of EAP in this area and can't be constrained because we won't document what we do. If EAP documenters want to reuse WildFly content in its docs, that's great, I'm all for that, but we should organize things so they can reasonably efficiently pick and choose what to use.

The "how to use things WildFly provides" part can also evolve over time; e.g. it's archetypes now because it's what we have, but later it can be a project generator or something.

Reply:
@bstansberry IMHO in WFLY we should have everything we have on EAP + unstable/upstream-only. We need to document the manual options upstream too.

Reply:
@bstansberry @jmesnil btw FYI this JIRA was taken over by DOCs team (this AD is from 2021), and the work is being led by Ashwin, with a minimal version for this to be available on EAP 8.0. Even if this is on docs domain now, we still have a perm status/progress discussion spot on BOMs and QS weekly.

Reply:
Understood, and we need to consider that downstream work, and I'm fine with orienting the scope of this proposal toward those requirements, as that will certainly be better than what we have now!

But WildFly needs to move beyond that, and quite soon, so it's valuable to consider where we want things to go.

Reply:
> are you sure it's the best direction for us to promote long term?

No. But this is our best direction for WildFly *short term*. We could have *now* a way for users to bootstrap their Entreprise Java applications in a matter of minutes. This would be a huge productivity boon.
Long term, there can be other solutions (code launcher) but there is no reason not to improve the user experience now with what we have now.
 
> Anyway let's say it takes 30 min for the user to manually add everything to the Maven Project... isn't that only for the 1st time for the user to learn about our plugins, BOMs, etc? And won't that learning be needed? And would actually those few minutes really mean something on the time it takes to develop a new app?

These few initial minutes are very important for the perception of the user and the developer joy. As a developer, do you want to spend 30 to 1 hour reading a doc and doing `mkdir`, `cp` all other the places before you are able to do your actual work? I definitely don't. I want to work on *my* stuff with a project that respects my time.

As a team, we agree that is one of our objective to be able to get a WildFly application up and running in less than 5 minutes (download times not included). This is a valuable objective that drives innovation in terms of user experience and feature capabilities.

> I can understand the need for an as fast as possible "server start", for an as fast as possible "get started", now an as fast as possible "app development"? I can't really see what's wrong with now explaining how to do it manually, starting with the maven archetypes + (upstream only) extend that with the "unstable/unsupported" features like Archetypes.

There is nothing wrong with "explaining how to do it manually". But from a developer experience, a "show (and tell)" user experience beats every time a "tell (and figure out what instructions you missed)" one :) 

> If you truly believe Archetypes is the direction we should go long term then why not an RFE and make them a product with a feature team?

I don't understand your sentence. We can release a new archetype now to improve WildFly user experience without committing to any product.
The important commitment is on the *output* of the tool: a Maven project properly setup to develop with WildFly.

And again, I don't think this is the ideal long term solution but this is something that would benefits our user now and do not prevent us to improve on it down the road...

Reply:
@emmartins I understand that this proposal is a downstream requirement for EAP 8. As @bstansberry suggested, let's make that a clear requirement and focus on it.

My only remark is then that this proposal should not hamper any progress or changes we need to improve WildFly user experience.

Reply:
> These few initial minutes are very important for the perception of the user and the developer joy. As a developer, do you want to spend 30 to 1 hour reading a doc and doing mkdir, cp all other the places before you are able to do your actual work? I definitely don't. I want to work on my stuff with a project that respects my time.

mkdir, cp, what are you talking about? I was just referring to what we have been using up till now, and what was being worked out by DOCs for this RFE up till now, which is the usage of the standard Maven archetypes. On top of using those you pretty much just add our plugin management, boms import, repos and that's is most cases wrt our stuff... Please note that WFLY Archetypes will never be enough for a real app, there is always user stuff to add to the Maven project (dependencies, repos, versions, etc.) so the time adding our stuff becomes irrelevant. 

> As a team, we agree that is one of our objective to be able to get a WildFly application up and running in less than 5 minutes (download times not included). This is a valuable objective that drives innovation in terms of user experience and feature capabilities.

I think we are too focused on the get started app and forget we are adding a new WFLY/EAP public component that may in the end may just be more noise on our and user's processes.

> I don't understand your sentence. We can release a new archetype now to improve WildFly user experience without committing to any product.

But we need to improve EAP user experience too... I kind of feel that perhaps I am not explaining my concerns properly but in short what is really in my mind is that we should spend some effort NOW figuring out what is the best long term solution, and work on it for both WFLY and EAP. And if we need to extend an OKR to another CY no prob.  In the end we may even end up deciding the archetypes are in fact the best direction!

Reply:
@bstansberry @jmesnil no matter the - very good IMHO - discussion the docs team let me know they will consider the  documentation of both archetypes, using a switch to render differently for WFLY and EAP.

Reply:
> I was just referring to what we have been using up till now, and what was being worked out by DOCs for this RFE up till now, which is the usage of the standard Maven archetypes.

Could you please link to this doc? I was not able to find it in https://docs.wildfly.org/29/Getting_Started_Developing_Applications_Guide.html

> I think we are too focused on the get started app and forget we are adding a new WFLY/EAP public component that may in the end may just be more noise on our and user's processes.

I never implied or suggested to create archetype for EAP. I'm focused on WildFly.
I'm not focused on the *archetype*. I'm focused on the Maven project setup. The archetype is what we have now (or close to have it) to produce this setup. Is there any other solution we provide to our users to bootstrap their applications *now*?

Reply:
@jmesnil You will find it in the READMEs of several MP Quickstarts, e.g. https://github.com/wildfly/quickstart/blob/a3b0be1576e2cdb93abdb8512074af5006f8e825/microprofile-config/README.adoc?plain=1#L56

Reply:
thanks (and that's exactly all the `cp` and `mkdir` that I was talking about to fill the general maven archetypes with the specifities of WildFly)

Reply:
Which mkdir or cp you referring to?

Reply:
all the xml snippets you have to copy to the pom.xml, the directories and files you create to populate `src`, etc.

Reply:
The only non-app-specific there are the pom.xml changes I previously referred, which have little time/effort impact due to the need for app-specific changes there too, e.g. dependencies.